beacon_node/beacon_chain/src/beacon_chain.rs

```diff
@@ -3218,6 +3218,14 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
         let slot = state.slot();
         let proposer_index = state.get_beacon_proposer_index(state.slot(), &self.spec)? as u64;
 
+        let pubkey_opt = match self.validator_pubkey_bytes(proposer_index as usize) {
```

Comment:
Since you have a `BeaconState` in scope here, I'd suggest doing `state.validators().get(proposer_index as usize).map(|v| v.pubkey)` rather than hitting the pubkey cache.

There can be some lock contention on the validator pubkey cache at times. There's also something very definite about saying "this validator doesn't exist in the state of the block they're proposing". You *could* even return an error in that case, I don't see how any block could be valid if the proposer index isn't in the validator registry. I would only do that if there's a benefit in dropping the `Option` around the `pubkey`, otherwise we might as well just leave it a warning and have one less way that block proposal can fail.

Reply:
Ok used this state to get the pubkey here: ac15207bf3e4a41239be1fc562297fb0c9916a68

I've left the `Option` because dropping it doesn't gain us anything, but that's a good point, if the validator isn't in the state I too have no idea how it could propose. Seemed more like a possibility when I was using the cache 😂